### ISSUE TITLE
Re: Self-signed certificates should include CN, it is not deprecated, only when used for verification

### ISSUE THREAD

**Issue by polarathene** [caddyserver/caddy#3755]
As @mholt locked [an issue with their last response](https://github.com/caddyserver/caddy/issues/3535), this issue serves as a followup response to relay [OWASP guidelines on the topic](https://cheatsheetseries.owasp.org/cheatsheets/Transport_Layer_Protection_Cheat_Sheet.html#use-correct-domain-names):

> Historically this was stored in the `commonName` (CN) attribute of the certificate. However, modern versions of Chrome ignore the CN attribute, and require that the FQDN is in the `subjectAlternativeName` (SAN) attribute. 
> For compatibility reasons, **certificates should have the primary FQDN in the CN**, and the full list of FQDNs in the SAN.

---

Meanwhile a resource such as [this one](https://support.dnsimple.com/articles/what-is-common-name/#common-name-vs-subject-alternative-name) highlight stance of @mholt 

> The CA/Browser Forum has since mandated that the SAN would also include any value present in the common name, effectively making the SAN the only required reference for a certificate match with the server name. 
> The notion of the common name survives mostly as a legacy of the past. There are active discussions to remove its use from most browsers and interfaces.
> Since HTTPS was first introduced in 2000 (and defined by the [RFC 2818](https://tools.ietf.org/html/rfc2818)), the use of the commonName field has been considered deprecated, because it’s ambiguous and untyped.

From RFC 2818 (May 2000):

> If a subjectAltName extension of type dNSName is present, that MUST be used as the identity. Otherwise, the (most specific) Common Name field in the Subject field of the certificate MUST be used. Although the use of the Common Name is existing practice, it is deprecated and Certification Authorities are encouraged to use the dNSName instead.

[Here is the Chrome announcment that references that RFC](https://www.chromestatus.com/feature/4981025180483584) for ignoring CN to match against SAN only.

[CA/B also have some guidelines](https://stackoverflow.com/a/29600674) which state for the CN attribute:

> Required/Optional: Deprecated (Discouraged, but not prohibited)

The StackOverflow answer link for that CA/B quote also mentions RFC 6125 from 2011 which further details SAN/CN behaviour, but **it should be noted that these RFCs are specifically regarding deprecation for verification purposes, not that CN attribute itself should not exist on a certificate.**

> However, the Common Name is not strongly typed because a Common Name might contain a human-friendly string for the service, rather than a string whose form matches that of a fully qualified DNS domain name - [RFC 6125](https://tools.ietf.org/html/rfc6125)

[LetsEncrypt Glossary](https://letsencrypt.org/docs/glossary/):

> Common Name (CN) : **Part of a certificate’s Subject describing what the certificate is about**. For roots and intermediates it’s the human-readable name of the certificate authority. For leaf certificates it’s one of the domain names on the certificate.
> It is an **obsolete method** of indicating a domain name to which the certificate applies, since current Internet standards expect software to check only the Subject Alternative Names in order to determine the applicability of a certificate.

In 2016 LetsEncrypt updated Boulder to [allow certs to have an empty CN](https://github.com/letsencrypt/boulder/issues/40), in 2018 [a LetsEncrypt engineer provided some additional insights](https://community.letsencrypt.org/t/how-to-obtain-a-cert-without-a-common-name/72807/4):

> My understanding is we can’t omit the CN all-together as it causes **pretty considerable compatibility issues** with a variety of client software.
> Our alternative was going to be to place something other than a domain identifier in the field, but a strict reading of the **baseline requirements forbids this**

[Another resource, has detailed history](https://frasertweedale.github.io/blog-redhat/posts/2017-07-11-cn-deprecation.html):

> **The use of the CN field to carry DNS names was never a standard.** The Common Name field does not have these semantics; but using the CN in this way was an approach that worked. This interpretation was later formalised by the CA/B Forum in their Baseline Requirements for CAs, but only as a reflection of a current practice in SSL/TLS server and client implementations. 
> Even in the **Baseline Requirements the CN was a second-class citizen**; they mandated that if the CN was present at all, it must reflect one of the DNSName or IP address values from the SAN extension. All public CAs had to comply with this requirement, which is why Chrome’s removal of CN support is only affecting private PKIs, not public web sites.

---

Note the age of that RFC, and that modern CAs such as LetsEncrypt still include a CN, such that Caddy websites like the Discourse forum have a CN in their certificate. Github does as well issued by DigiCert.

So given all the above, [@mholt statement](https://github.com/caddyserver/caddy/issues/3535#issuecomment-699188937) seems invalid?:

> Using the CommonName field after it has been deprecated is *incorrect* software;

Both Firefox and Chrome correctly use the CN for UI inspection of a certificate. The fact that neither browser uses the CN for verification and will use SAN instead, should only further emphasize the validity of CN attribute still being relevant to a certificate for other purposes, the browsers clearly acknowledge this, as does additional resources for clarifying that above I hope?

@aduzsardi and myself both see value in the CN from a UX standpoint, and no negative consequence for including it for this purpose (it's ignored / deprecated for cert validation, so non-issue there).

## TL;DR

Hence, Caddy is *incorrect*, it has misunderstood what is deprecated. Common Name attribute is only deprecated for the purpose of certificate verification, the attribute still serves a valid non-deprecated purpose for administration/inspection.

**Comment by aduzsardi** [comment]
nicely done @polarathene , kudos for the research done on the matter
